PWGHF: Add OmegaC0 workflows that use KFParticle

Yunfan-Liu opened the pull request:
- add candidateCreator, candidateSelector, and treeCreator for omegac0 -> omega pi with KFParticle.
- add a machine learning header file, which is used in candidateSelector.

vkucera commented on PWGHF/DataModel/CandidateReconstructionTables.h:
```diff
@@ -1902,6 +1902,253 @@ DECLARE_SOA_TABLE(HfCandDstarMcGen, "AOD", "HFCANDDSTRMCGEN",
+                  hf_cand_omegac::FlagMcMatchGen, hf_cand_omegac::DebugGenCharmBar, hf_cand_omegac::DebugGenOmega, hf_cand_omegac::DebugGenLambda, hf_cand_omegac::PtCharmBaryonGen, hf_cand_omegac::EtaCharmBaryonGen, hf_cand_omegac::OriginGen);
+
+// table with results of KFParticle
+DECLARE_SOA_TABLE(HfOmegaCKF, "AOD", "HFOMEGACKF", //!
```
```suggestion
DECLARE_SOA_TABLE(HfOmegacKf, "AOD", "HFOMEGACKF", //!
```

vkucera commented on PWGHF/TableProducer/candidateCreatorOmegacToOmegaPiWithKFP.cxx:
```diff
@@ -0,0 +1,1119 @@
+  Configurable<bool> kfDoDCAFitterPreMinimCasc{"kfDoDCAFitterPreMinimCasc", false, "KF: do DCAFitter pre-optimization before KF fit to include material corrections for Omega"};
+  Configurable<bool> kfDoDCAFitterPreMinimCharmBaryon{"kfDoDCAFitterPreMinimCharmBaryon", true, "KF: do DCAFitter pre-optimization before KF fit to include material corrections for OmegaC0"};
+  // hist
+  HistogramRegistry registry{
+    "registry",
+    {
+
+    }};
```
One line please.

vkucera commented on PWGHF/TableProducer/candidateSelectorOmegacToOmegaPiWithKFP.cxx:
```diff
@@ -0,0 +1,718 @@
+#include "PWGHF/DataModel/CandidateReconstructionTables.h"
+#include "PWGHF/DataModel/CandidateSelectionTables.h"
+
+using namespace o2;
+using namespace o2::aod;
+using namespace o2::framework;
+using namespace o2::analysis;
+int toCounter = 0;
+enum pidInfoStored {
+  kPiFromLam = 0,
+  kPrFromLam,
+  kKaFromCasc,
+  kPiFromCharm
+};
```
Keep proper spacing between blocks.

vkucera commented on PWGHF/TableProducer/candidateSelectorOmegacToOmegaPiWithKFP.cxx:
```diff
@@ -0,0 +1,718 @@
+using namespace o2::aod;
+using namespace o2::framework;
+using namespace o2::analysis;
+int toCounter = 0;
```
What does it do?

vkucera commented on PWGHF/TableProducer/candidateSelectorOmegacToOmegaPiWithKFP.cxx:
```diff
@@ -0,0 +1,718 @@
+  Produces<aod::HfMlSelOmegacToOmegaPi> hfMlToOmegaPiCandidate;
+
+  // LF analysis selections
+  Configurable<double> radiusCascMin{"radiusCascMin", 0.6, "Min cascade radius"};
+  Configurable<double> radiusV0Min{"radiusV0Min", 1.2, "Min V0 radius"};
+  Configurable<double> cosPAV0Min{"cosPAV0Min", 0.97, "Min valueCosPA V0"};
+  Configurable<double> cosPACascMin{"cosPACascMin", 0.97, "Min value CosPA cascade"};
+  Configurable<double> dcaCascDauMax{"dcaCascDauMax", 1.0, "Max DCA cascade daughters"};
+  Configurable<double> dcaV0DauMax{"dcaV0DauMax", 1.0, "Max DCA V0 daughters"};
```
Why `double`?

vkucera commented on PWGHF/TableProducer/candidateSelectorOmegacToOmegaPiWithKFP.cxx:
```diff
@@ -0,0 +1,718 @@
+  // using TracksSel = soa::Join<aod::TracksWDcaExtra, aod::TracksPidPi, aod::TracksPidPr, aod::TracksPidKa, aod::PidTpcTofFullPi, aod::PidTpcTofFullPr, aod::PidTpcTofFullKa>;
+  HistogramRegistry registry{"registry"}; // for QA of selections
+
+  OutputObj<TH1F> hInvMassCharmBaryon{TH1F("hInvMassCharmBaryon", "Charm baryon invariant mass;inv mass;entries", 500, 2.2, 3.1)};
```
Use `registry`

vkucera commented on PWGHF/TableProducer/candidateSelectorOmegacToOmegaPiWithKFP.cxx:
```diff
@@ -0,0 +1,718 @@
+               TracksSel const&)
+  {
+
+    double massLambdaFromPDG = o2::constants::physics::MassLambda0;
+    double massOmegaFromPDG = o2::constants::physics::MassOmegaMinus;
```
Remove

vkucera commented on PWGHF/TableProducer/treeCreatorOmegacToOmegaPiWithKFP.cxx:
```diff
@@ -0,0 +1,790 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file
+/// \brief Writer of the omegac0  to Omega Pi candidates in the form of flat tables to be stored in TTrees.
+///        In this file are defined and filled the output tables
+///
+/// \author
```
Complete

vkucera commented on PWGHF/TableProducer/treeCreatorOmegacToOmegaPiWithKFP.cxx:
```diff
@@ -0,0 +1,790 @@
+  Produces<o2::aod::HfOmegacLites> rowCandidateLite;
+  Produces<o2::aod::HfKFOmegacFulls> rowKFCandidateFull;
+  Produces<o2::aod::HfMlKFOmegacFulls> rowMlKFCandidateFull;
+  Produces<o2::aod::HfOmegacEvs> rowEv;
+  Configurable<float> zPvCut{"zPvCut", 10., "Cut on absolute value of primary vertex z coordinate"};
```
```suggestion
  Produces<o2::aod::HfOmegacEvs> rowEv;

  Configurable<float> zPvCut{"zPvCut", 10., "Cut on absolute value of primary vertex z coordinate"};
```

fgrosa reviewed (changes requested):
@Yunfan-Liu thanks for the implementation! On top of the comments of @vkucera I have a more general comment. The selector, creator, and tree creator of the `OmegaCToOmegaPi` channel already exist:

- https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx
- https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateSelectorOmegac0ToOmegaPi.cxx
- https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/treeCreatorOmegacToOmegaPi.cxx

In order to avoid code duplication it would be better to just implement the option to use the KFParticle in them (as for the D0). This also goes in the direction of re-using what the columns that are already implemented for these channels, and just add the KF-related ones.

Yunfan-Liu commented on PWGHF/TableProducer/candidateCreatorOmegacToOmegaPiWithKFP.cxx:
```diff
@@ -0,0 +1,1119 @@
+  Configurable<double> maxDXYIni{"maxDXYIni", 4., "reject (if>0) PCA candidate if tracks DXY exceeds threshold"};
+  Configurable<double> minParamChange{"minParamChange", 1.e-3, "stop iterations if largest change of any X is smaller than this"};
+  Configurable<double> minRelChi2Change{"minRelChi2Change", 0.9, "stop iterations is chi2/chi2old > this"};
+  Configurable<double> maxChi2{"maxChi2", 100., "discard vertices with chi2/Nprongs > this (or sum{DCAi^2}/Nprongs for abs. distance minimization)"};
```
> Why do you support DCA fitter in a workflow that is dedicated to KF particle?

I have seen DCAfitter as an option in casadebuilder in KF process function, and now I am sure DCAfitter is unnecessary. I delete DCAfitter from this workflow.

Yunfan-Liu commented:
> @Yunfan-Liu thanks for the implementation! On top of the comments of @vkucera I have a more general comment. The selector, creator, and tree creator of the `OmegaCToOmegaPi` channel already exist:
> 
> * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx
> * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateSelectorOmegac0ToOmegaPi.cxx
> * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/treeCreatorOmegacToOmegaPi.cxx
> 
> In order to avoid code duplication it would be better to just implement the option to use the KFParticle in them (as for the D0). This also goes in the direction of re-using what the columns that are already implemented for these channels, and just add the KF-related ones.

Thanks for your suggestion. I have not yet had any contact or discussion with them, so I want to first submit an independent KFparticle workflow. Later, I will find a way to add KFparticle as an option to the existing workflow.

fgrosa commented:
> > @Yunfan-Liu thanks for the implementation! On top of the comments of @vkucera I have a more general comment. The selector, creator, and tree creator of the `OmegaCToOmegaPi` channel already exist:
> > 
> > * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx
> > * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateSelectorOmegac0ToOmegaPi.cxx
> > * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/treeCreatorOmegacToOmegaPi.cxx
> > 
> > In order to avoid code duplication it would be better to just implement the option to use the KFParticle in them (as for the D0). This also goes in the direction of re-using what the columns that are already implemented for these channels, and just add the KF-related ones.
> 
> @fgrosa Thanks for your suggestion. I have not yet had any contact or discussion with them, so I want to first submit an independent KFparticle workflow. Later, I will find a way to add KFparticle as an option to the existing workflow.

Hi @Yunfan-Liu thanks for your reply. However, it would be very good to do it for the first version instead of duplicating a lot of code, because otherwise it will become much more difficult in the future, when we risk that the codes diverge. I also think that there is in principle no need to discuss with the other analysers because, unless we introduce bugs, the additions should be transparent for the other option (with DCAFitter). Thanks for your understanding!

Yunfan-Liu commented:
> > > @Yunfan-Liu thanks for the implementation! On top of the comments of @vkucera I have a more general comment. The selector, creator, and tree creator of the `OmegaCToOmegaPi` channel already exist:
> > > 
> > > * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx
> > > * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateSelectorOmegac0ToOmegaPi.cxx
> > > * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/treeCreatorOmegacToOmegaPi.cxx
> > > 
> > > In order to avoid code duplication it would be better to just implement the option to use the KFParticle in them (as for the D0). This also goes in the direction of re-using what the columns that are already implemented for these channels, and just add the KF-related ones.
> > 
> > 
> > @fgrosa Thanks for your suggestion. I have not yet had any contact or discussion with them, so I want to first submit an independent KFparticle workflow. Later, I will find a way to add KFparticle as an option to the existing workflow.
> 
> Hi @Yunfan-Liu thanks for your reply. However, it would be very good to do it for the first version instead of duplicating a lot of code, because otherwise it will become much more difficult in the future, when we risk that the codes diverge. I also think that there is in principle no need to discuss with the other analysers because, unless we introduce bugs, the additions should be transparent for the other option (with DCAFitter). Thanks for your understanding!

@fgrosa Thanks for your reply, I will follow your suggestion. Now I am trying to implement the KFParticle as an option in the existing workflow.

vkucera commented on PWGHF/DataModel/CandidateReconstructionTables.h:
```diff
@@ -1902,6 +1902,253 @@ DECLARE_SOA_TABLE(HfCandDstarMcGen, "AOD", "HFCANDDSTRMCGEN",
+                  hf_cand_omegac::CollisionMatched);
+
+// table with results of generator level MC matching
+DECLARE_SOA_TABLE(HfOmegaCMCGen, "AOD", "HFOMEGACMCGEN", //!
```
`MC` -> `Mc`

Yunfan-Liu commented:
> > > @Yunfan-Liu thanks for the implementation! On top of the comments of @vkucera I have a more general comment. The selector, creator, and tree creator of the `OmegaCToOmegaPi` channel already exist:
> > > 
> > > * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx
> > > * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/candidateSelectorOmegac0ToOmegaPi.cxx
> > > * https://github.com/AliceO2Group/O2Physics/blob/master/PWGHF/TableProducer/treeCreatorOmegacToOmegaPi.cxx
> > > 
> > > In order to avoid code duplication it would be better to just implement the option to use the KFParticle in them (as for the D0). This also goes in the direction of re-using what the columns that are already implemented for these channels, and just add the KF-related ones.
> > 
> > 
> > @fgrosa Thanks for your suggestion. I have not yet had any contact or discussion with them, so I want to first submit an independent KFparticle workflow. Later, I will find a way to add KFparticle as an option to the existing workflow.
> 
> Hi @Yunfan-Liu thanks for your reply. However, it would be very good to do it for the first version instead of duplicating a lot of code, because otherwise it will become much more difficult in the future, when we risk that the codes diverge. I also think that there is in principle no need to discuss with the other analysers because, unless we introduce bugs, the additions should be transparent for the other option (with DCAFitter). Thanks for your understanding!

@fgrosa @vkucera, I apologize for taking up so much of your time. Now, following @fgrosa suggestion, I have added KFParticle as an option to the existing workflow, so I have deleted the workflow I created before.

fgrosa commented on PWGHF/DataModel/CandidateReconstructionTables.h:
```diff
@@ -1164,6 +1164,59 @@ DECLARE_SOA_COLUMN(DecLenV0, decLenV0, float);
+DECLARE_SOA_COLUMN(DecayLenXYLambda, decayLenXYLambda, float);
+DECLARE_SOA_COLUMN(DecayLenXYCasc, decayLenXYCasc, float);
+DECLARE_SOA_COLUMN(DecayLenXYOmegac, decayLenXYOmegac, float);
+DECLARE_SOA_COLUMN(CosPaV0ToCasc, cosPaV0ToCasc, float);
+DECLARE_SOA_COLUMN(CosPaV0ToPv, cosPaV0ToPv, float);
+DECLARE_SOA_COLUMN(CosPaCascToOmegac, cosPaCascToOmegac, float);
+DECLARE_SOA_COLUMN(CosPaCascToPv, cosPaCascToPv, float);
+DECLARE_SOA_COLUMN(CosPaOmegacToPv, cosPaOmegacToPv, float);
```
The Cosp variables are already defined for the non-KF case, please use the already defined ones (unless there is any other reason to duplicate them)

Yunfan-Liu replied:
Thanks. I deleted them and used the existing column.

fgrosa commented on PWGHF/DataModel/CandidateReconstructionTables.h:
```diff
@@ -1164,6 +1164,59 @@ DECLARE_SOA_COLUMN(DecLenV0, decLenV0, float);
+DECLARE_SOA_COLUMN(KfptPiFromOmegac, kfptPiFromOmegac, float);
+DECLARE_SOA_COLUMN(KfptOmegac, kfptOmegac, float);
+DECLARE_SOA_COLUMN(CosThetaStarPiFromOmegac, cosThetaStarPiFromOmegac, float);
+DECLARE_SOA_COLUMN(CtOmegac, ctOmegac, float);
+DECLARE_SOA_COLUMN(EtaOmegac, etaOmegac, float);
```
Ct and Eta are already defined for the non-KF case, please use them

Yunfan-Liu replied:
Done.

fgrosa commented on PWGHF/DataModel/CandidateReconstructionTables.h:
```diff
@@ -1272,6 +1344,30 @@ DECLARE_SOA_TABLE(HfCandToOmegaK, "AOD", "HFCANDTOOMEGAK",
                   o2::soa::Marker<2>);
 
+// table with results of KFParticle
+DECLARE_SOA_TABLE(HfOmegacKf, "AOD", "HFOMEGACKF", //!
```
It is very good to have a separated table for the KF variables. I was wondering though if you need to duplicate some of them, for instance the cosine of pointing angle: is it needed to be there both in the "normal" and KF particle tables? Can't you simply replace the normal cosp with the one computed with the KF, when the KFVertexing is used in place of the DCAFitter?

My suggestion would be to keep the original table as it is, filling it with variables computed with DCAFitter or KFParticle (depending on what is chosen by the user) and having this additional table only for KFParticle specific variables, that do not exist at all with the DCAFitter (for instance all the `Chi2Geo` and `Chi2Topo`). This is what was done for the D0, where only a small additional table is filled in case of KF, see https://github.com/AliceO2Group/O2Physics/blob/b2c8a33a4e8f959849aeb1c2eea40312ba8f4e27/PWGHF/TableProducer/candidateCreator2Prong.cxx#L430-L444

Yunfan-Liu replied:
Thanks for your suggestion. I delete the duplication from the KF table. The variables I want to fill in the KF table with are those output in the "Omegac0 to Omega Pi" task in Run 2. 
https://github.com/alisw/AliPhysics/blob/8f7fdfc00228ff5f41b02305b26f161534517d48/PWGHF/vertexingHF/AliAnalysisTaskSEXicZero2XiPifromKFP.cxx#L4879-L4931
I think these variables might be useful for machine learning, so I want to save them locally.

fgrosa replied:
Thanks! I think you still have these ones that are duplicated:
```
hf_cand_xic0_omegac0::CosPaV0ToPv, hf_cand_xic0_omegac0::CosPaCascToPv,
```

fgrosa replied:
The variables you want to store locally should be defined in the tree creator, and can be either in one or the other table produced by the candidate creator.

Yunfan-Liu replied:
Thanks for your suggestion. It's done.

fgrosa commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -489,6 +574,440 @@ struct HfCandidateCreatorXic0Omegac0 {
+                       kfOmegac0Candidate.kfDcaCascDau, kfOmegac0Candidate.kfDcaV0Dau, kfOmegac0Candidate.kfDcaOmegacDau,
+                       decLenCharmBaryon, decLenCascade, decLenV0, errorDecayLengthCharmBaryon, errorDecayLengthXYCharmBaryon);
+      // fill kf table
+      kfCandidateData(kfOmegac0Candidate.nSigmaTPCPiFromOmegac, kfOmegac0Candidate.nSigmaTOFPiFromOmegac,
+                      kfOmegac0Candidate.nSigmaTOFKaFromCasc, kfOmegac0Candidate.nSigmaTOFKaFromCasc,
+                      kfOmegac0Candidate.nSigmaTPCPiFromV0, kfOmegac0Candidate.nSigmaTPCPrFromV0,
+                      kfOmegac0Candidate.kfDcaXYPiFromOmegac, kfOmegac0Candidate.kfDcaCascDau, kfOmegac0Candidate.kfDcaOmegacDau, kfOmegac0Candidate.kfDcaXYCascToPv,
+                      kfOmegac0Candidate.chi2GeoV0, kfOmegac0Candidate.chi2GeoCasc, kfOmegac0Candidate.chi2GeoOmegac, kfOmegac0Candidate.chi2MassV0, kfOmegac0Candidate.chi2MassCasc,
+                      kfOmegac0Candidate.ldlV0, kfOmegac0Candidate.ldlCasc, kfOmegac0Candidate.ldlOmegac,
+                      kfOmegac0Candidate.chi2TopoV0ToPv, kfOmegac0Candidate.chi2TopoCascToPv, kfOmegac0Candidate.chi2TopoPiFromOmegacToPv, kfOmegac0Candidate.chi2TopoOmegacToPv,
+                      kfOmegac0Candidate.chi2TopoV0ToCasc, kfOmegac0Candidate.chi2TopoCascToOmegac,
+                      kfOmegac0Candidate.decayLenXYLambda, kfOmegac0Candidate.decayLenXYCasc, kfOmegac0Candidate.decayLenXYOmegac,
+                      kfOmegac0Candidate.cosPaV0ToCasc, kfOmegac0Candidate.cosPaV0ToPv, kfOmegac0Candidate.cosPaCascToOmegac, kfOmegac0Candidate.cosPaCascToPv, kfOmegac0Candidate.cosPaOmegacToPv,
+                      kfOmegac0Candidate.massV0, kfOmegac0Candidate.massCasc, kfOmegac0Candidate.massOmegac,
+                      kfOmegac0Candidate.rapOmegac, kfOmegac0Candidate.ptPiFromOmegac, kfOmegac0Candidate.ptOmegac,
+                      kfOmegac0Candidate.cosThetaStarPiFromOmegac,
+                      kfOmegac0Candidate.ctOmegac,
+                      kfOmegac0Candidate.etaOmegac,
+                      v0NDF, cascNDF, charmbaryonNDF, v0NDF_m, cascNDF_m,
+                      v0Chi2OverNdf, cascChi2OverNdf, charmbaryonChi2OverNdf, v0Chi2OverNdf_m, cascChi2OverNdf_m);
```
You have several variables that are duplicated in the two tables, such as `kfOmegac0Candidate.cosPaV0ToCasc`. Please do not duplicate them (they increase the memory for no reason), see also my comment above.

Yunfan-Liu replied:
Done.

fgrosa reviewed (changes requested):
Hi @Yunfan-Liu thank you very much for considering my request, the code looks much better now already. I still have a few comments about the variables and mainly their duplication in some cases before merging.

Yunfan-Liu commented on PWGHF/Core/HfMlResponseOmegacToOmegaPi.h:
```diff
@@ -0,0 +1,433 @@
+  decLenV0,
+  errorDecayLengthCharmBaryon,
+  errorDecayLengthXYCharmBaryon,
+  ptCharmBaryon,
```
I deleted HfMlResponseOmegacToOmegaPi.h. I'm now trying to follow the omegac0 implementation of machine learning in Run 2 and make a first attempt in Run 3. I haven't actually identified which variables will be used in machine learning yet, so I've added all the variables that might be used to the header.
I think it would be better to add the machine learning header after analysing it locally and identifying the variables I need, so I'm deleting it now.

Yunfan-Liu commented on PWGHF/DataModel/CandidateReconstructionTables.h:
```diff
@@ -1164,6 +1164,59 @@ DECLARE_SOA_COLUMN(DecLenV0, decLenV0, float);
 DECLARE_SOA_COLUMN(ErrorDecayLengthXYCharmBaryon, errorDecayLengthXYCharmBaryon, float);
 
+// KFParticle results
+DECLARE_SOA_COLUMN(NSigmaTPCPiFromOmegac, nSigmaTPCPiFromOmegac, float);
+DECLARE_SOA_COLUMN(NSigmaTOFPiFromOmegac, nSigmaTOFPiFromOmegac, float);
+DECLARE_SOA_COLUMN(NSigmaTPCKaFromCasc, nSigmaTPCKaFromCasc, float);
+DECLARE_SOA_COLUMN(NSigmaTOFKaFromCasc, nSigmaTOFKaFromCasc, float);
+DECLARE_SOA_COLUMN(NSigmaTPCPiFromV0, nSigmaTPCPiFromV0, float);
+DECLARE_SOA_COLUMN(NSigmaTPCPrFromV0, nSigmaTPCPrFromV0, float);
```
I saw that these variables were used in pT-dependent pre-selections in the Run2 analysis, so I want to store them and output them locally.

fgrosa replied:
This you can do with the tree creator, there is no need to have them in the table of the candidate creator

Yunfan-Liu replied:
Thanks for your suggestion. It's done.

Yunfan-Liu commented on PWGHF/DataModel/CandidateReconstructionTables.h:
```diff
@@ -1164,6 +1164,59 @@ DECLARE_SOA_COLUMN(DecLenV0, decLenV0, float);
+DECLARE_SOA_COLUMN(Chi2TopoOmegacToPv, chi2TopoOmegacToPv, float);
+DECLARE_SOA_COLUMN(Chi2TopoV0ToCasc, chi2TopoV0ToCasc, float);
+DECLARE_SOA_COLUMN(Chi2TopoCascToOmegac, chi2TopoCascToOmegac, float);
+DECLARE_SOA_COLUMN(DecayLenXYLambda, decayLenXYLambda, float);
+DECLARE_SOA_COLUMN(DecayLenXYCasc, decayLenXYCasc, float);
+DECLARE_SOA_COLUMN(DecayLenXYOmegac, decayLenXYOmegac, float);
```
These decay lengths are obtained from the KF package.

Yunfan-Liu commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -489,6 +574,437 @@ struct HfCandidateCreatorXic0Omegac0 {
+                       v0index, casc.posTrackId(), casc.negTrackId(),
+                       casc.cascadeId(), trackCharmBachelor.globalIndex(), casc.bachelorId(),
+                       kfOmegac0Candidate.massV0, kfOmegac0Candidate.massCasc, kfOmegac0Candidate.massOmegac,
+                       kfOmegac0Candidate.cosPaV0ToCasc, kfOmegac0Candidate.cosPaOmegacToPv, kfOmegac0Candidate.cosPaCascToOmegac, kfOmegac0Candidate.cosPaXYV0ToCasc, kfOmegac0Candidate.cosPaXYOmegacToPv, kfOmegac0Candidate.cosPaXYCascToOmegac,
```
Thanks for the correction. It's done.

fgrosa commented:
Hi @Yunfan-Liu thanks a lot for considering my comments! I noticed that you deleted the class for the ML inference, was it intentional? If so, I am fine with the current version and, if good also for @vkucera, we can merge it.

Yunfan-Liu commented:
> Hi @Yunfan-Liu thanks a lot for considering my comments! I noticed that you deleted the class for the ML inference, was it intentional? If so, I am fine with the current version and, if good also for @vkucera, we can merge it.

Yes. now I am unsure what variables will be used in ML, so I deleted it.

fgrosa commented:
> > Hi @Yunfan-Liu thanks a lot for considering my comments! I noticed that you deleted the class for the ML inference, was it intentional? If so, I am fine with the current version and, if good also for @vkucera, we can merge it.
> 
> Yes. now I am unsure what variables will be used in ML, so I deleted it.

Approved from my side, when also Vit approves it can be merged.

alibuild commented:
Error while checking build/O2Physics/o2 for 64626cddd3ab6285a91f5634966e5fd405d89a29 at 2024-07-11 19:33:
```
## sw/BUILD/O2Physics-latest/log
/sw/SOURCES/O2Physics/6327-slc7_x86-64/0/PWGHF/TableProducer/treeCreatorOmegacToOmegaPi.cxx:297:66: error: unused parameter 'debugMc' [-Werror=unused-parameter]
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/6327/64626cddd3ab6285a91f5634966e5fd405d89a29/build_O2Physics_o2/pretty.html).

fgrosa commented:
Hi @vkucera could you please check if this PR (which is stalled since a while) is good for you? Thanks a lot!

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -35,13 +35,21 @@
 
 #include "PWGLF/DataModel/LFStrangenessTables.h"
 
+#include "Tools/KFparticle/KFUtilities.h"
```
This is a general O2Physics header. It should go after `Common/...`

Yunfan-Liu replied:
Done.

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -35,13 +35,21 @@
 #include "PWGHF/Utils/utilsBfieldCCDB.h"
 #include "PWGHF/Utils/utilsEvSelHf.h"
 
+/// includes KFParticle
+#include "KFParticle.h"
+#include "KFParticleBase.h"
+#include "KFPTrack.h"
+#include "KFPVertex.h"
+#include "KFVertex.h"
```
These are external headers. They should be included before O2 headers.

Yunfan-Liu replied:
Done.

vkucera reviewed (changes requested):
@Yunfan-Liu thanks for addressing the comments. The PR looks good to me after fixing the remaining minor issues.

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -13,6 +13,13 @@
 /// \brief Reconstruction of Omegac0 and Xic0 decays candidates
 /// \author Federica Zanone <federica.zanone@cern.ch>, Heidelberg University
```
Please add yourself on the list of authors with a second `/// \author` line.

Yunfan-Liu replied:
Done.

vkucera commented on PWGHF/TableProducer/treeCreatorOmegacToOmegaPi.cxx:
```diff
@@ -99,6 +99,53 @@ DECLARE_SOA_COLUMN(TofNSigmaPiFromCharmBaryon, tofNSigmaPiFromCharmBaryon, float
 DECLARE_SOA_COLUMN(TofNSigmaKaFromCasc, tofNSigmaKaFromCasc, float);
 DECLARE_SOA_COLUMN(TofNSigmaPiFromLambda, tofNSigmaPiFromLambda, float);
 DECLARE_SOA_COLUMN(TofNSigmaPrFromLambda, tofNSigmaPrFromLambda, float);
+// from creator KF
```
Please add yourself on the list of authors with a second `/// \author` line.

alibuild commented:
Error while checking build/O2Physics/o2 for 4d023e49d9a460166c2102a1db063b14af5e33b3 at 2024-07-30 19:24:
```
## sw/BUILD/O2Physics-latest/log
/sw/SOURCES/O2Physics/6327-slc7_x86-64/0/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:600:19: error: 'SetField' is not a member of 'KFParticle'
/sw/SOURCES/O2Physics/6327-slc7_x86-64/0/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:576:72: error: unused parameter 'bcWithTimeStamps' [-Werror=unused-parameter]
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/6327/4d023e49d9a460166c2102a1db063b14af5e33b3/build_O2Physics_o2/pretty.html).

vkucera commented on PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:
```diff
@@ -42,6 +39,13 @@
 #include "Common/DataModel/EventSelection.h"
 #include "Tools/KFparticle/KFUtilities.h"
 
+/// includes KFParticle
+#include "KFParticle.h"
+#include "KFParticleBase.h"
+#include "KFPTrack.h"
+#include "KFPVertex.h"
+#include "KFVertex.h"
```
Why did you move the headers?

Yunfan-Liu replied:
Sorry, that's a mistake. I moved these headers back to their original positions. 
I just found that if I add
``` 
#ifndef HomogeneousField
#define HomogeneousField
#endif
  ```
to the front of these headers, this error is resolved
> Error while checking build/O2Physics/o2 for [4d023e4](https://github.com/AliceO2Group/O2Physics/commit/4d023e49d9a460166c2102a1db063b14af5e33b3) at 2024-07-30 19:24:
> 
> ```
> ## sw/BUILD/O2Physics-latest/log
> /sw/SOURCES/O2Physics/6327-slc7_x86-64/0/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:600:19: error: 'SetField' is not a member of 'KFParticle'
> /sw/SOURCES/O2Physics/6327-slc7_x86-64/0/PWGHF/TableProducer/candidateCreatorXic0Omegac0.cxx:576:72: error: unused parameter 'bcWithTimeStamps' [-Werror=unused-parameter]
> ninja: build stopped: subcommand failed.
> ```
> 
> Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/O2Physics/6327/4d023e49d9a460166c2102a1db063b14af5e33b3/build_O2Physics_o2/pretty.html).
